### ISSUE TITLE
Why is `this.path` set to undefined instead of `path`?

### ISSUE THREAD

**Issue by IamLizu** [pillarjs/router#122]
Can someone please help me understand why `this.path` is set to `undefined`?
https://github.com/pillarjs/router/blob/2e7fb67ad1b0c1cd2d9eb35c2244439c5c57891a/lib/layer.js#L43

I had checked the router _Layer_, and the path is was `undefined` in almost every route, that shouldn't be the case, right?

While I don't know the reason of it being set to `undefined` initially, I can suggest it should be set to `path`. I have noticed that the `path` on _Layer_ is no longer undefined if we set,
```js
this.path = path;
```

**Comment by IamLizu** [comment]
Related https://github.com/expressjs/express/discussions/5961

**Comment by carpasse** [comment]
It seems that `this.path` is being set to `undefined` because we can't determine what `path` the `Layer` is matched against until runtime.

Let me explain my thought process: Layers are added to the Router instance when the [`use`](https://github.com/pillarjs/router/blob/2e7fb67ad1b0c1cd2d9eb35c2244439c5c57891a/index.js#L391) and [`route`](https://github.com/pillarjs/router/blob/2e7fb67ad1b0c1cd2d9eb35c2244439c5c57891a/index.js#L421) methods are executed.

The `Router.prototype.run` method only accepts a string as `path`, but `Router.prototype.use` can accept either a string or an array of strings. You can refer to the Express [documentation](https://expressjs.com/en/4x/api.html#app.use) for more details.

We can't know which of the possible `path` values was matched until the layer is actually matched at runtime, as seen [here](https://github.com/pillarjs/router/blob/2e7fb67ad1b0c1cd2d9eb35c2244439c5c57891a/index.js#L234).

To address the issue discussed in [express#5961](https://github.com/expressjs/express/discussions/5961), perhaps we could add an extra property to the Layer instance, such as `paths` or `pathPatterns`. This would make it possible to list all routes or a router.

@pillarjs/express-tc, can we get your input on this matter?

**Comment by IamLizu** [comment]
Thank you for the explanation @carpasse.

I still have some confusion, for example, the actual matching against a request's path happens at runtime. However, if we set the `Layer`'s `path` property when the `Layer` is created, will it create any issue?

**Comment by carpasse** [comment]
@IamLizu I believe the `path` argument passed to the `Layer` constructor represents either a list of handler paths (as an array) or a single string. To improve clarity, It may be best to rename that constructor parameter to something like `pathPatterns` or `layerPaths`, as I think it better reflects the purpose.

* [Reference 1](https://github.com/pillarjs/router/blob/master/lib/layer.js#L32) - Layer constructor
* [Reference 2](https://github.com/pillarjs/router/blob/master/index.js#L391C29-L391C33) - Layer instance creation

Additionally, the `path` property is expected to hold the matched portion of the runtime path based on the patterns passed to the constructor. However, this "path" is not one of the original paths passed in, but rather a substring derived from the matched path during runtime.

* This `path` is a substring match of the runtime path, not directly one of the paths passed to the constructor. [Reference 3](https://github.com/pillarjs/router/blob/master/lib/layer.js#L78)
* The `path` property cannot be defined until it has been matched against a runtime path. [Reference 4](https://github.com/pillarjs/router/blob/master/index.js#L234)
* After the match, it is effectively defined as `layerPath`, and then used to trim off the matched portion from the runtime path. [Reference 5](https://github.com/pillarjs/router/blob/master/index.js#L320)

**Comment by wesleytodd** [comment]
Yep pretty much what @carpasse has said. One thing I might ask is "what is the goal of the question"? Did you have an outcome you are looking to achieve, or is this more about exploration and learning? If the goal is listing all the routes (as is referenced near the top) then there are a bunch of discussions on that that might provide more context. But before spending time (I still have nearly 300 issues I am trying to catch up on), can I ask for a more clear explanation of the goal?

**Comment by IamLizu** [comment]
Thanks @carpasse for the explanation. 

@wesleytodd actually the outcome is nothing else than listing the routes. It appears setting the `path` to `this.path` instead of `undefined` helps easily listing the routes.

**Comment by wesleytodd** [comment]
Ah yes, one of my goals over the next weeks is to update that open api package. I was just slammed with other things through the end of last year. @mxngls if you are interested in helping with that work I might be able to give a few pointers. Honestly I think express/router itself should have some better apis for this kind of thing so that third parties dont have so much work to do for compatibility in as we make future changes. So might be we could work on landing that kind of stuff here as well to help.

**Comment by mxngls** [comment]
Thank you for your swift response @wesleytodd. Actually I already came up with a solution, that works quite well. I might have time to publish the refactored package over the weekend. I think to get rid of regular expressions for paths was a good decision overall. The work that was necessary to correctly parse these paths with regular expressions seems incredible convoluted and quite error prone. 

I intend to make the first version only compatible to Express version 5 and upwards for the same reasons. What is still left to do is to transform the paths as used by Express to something that is accurate to the OpenAPI specification. I don't think this will prove to be too cumbersome, as we only need to adjust the syntax for regular and optional parameters. Wildcard parameters are not conform to the spec anyway. I would love to ping you again and if possible get some input from your side. The package itself is written in Typescript and considerable leaner than the original _express-openapi_ package written in Javascript. But I think it would be better to move that talk into a separate discussion as it's not really an issue.

Lastly I fully agree regarding your comments on adding some proper APIs to Express so that obtaining router paths is not as arduous as it is right now.

**Comment by hanse** [comment]
This is kind of blocking us from upgrading from express v4 and express v5 as we rely on being able to infer the full list of routes for e.g. populating a openapi spec or listing all available endpoints. Our express v4 solution works perfectly well, but it does not work fully in express v5 and it seems partly due to internal changes in `router` making in impossible(?) to infer full paths for nested routers.

```typescript
router = Router()
router.get("/b", () => {})

app.use("/a", router)
```

**Comment by mxngls** [comment]
@wesleytodd Took me way too long to finally get this on GitHub but here we are: [express-openapi](https://github.com/mxngls/express-openapi) (I decided to keep the name as the old version for now). We've been using this in production for a while now and it works pretty well. I am sure it has some rough edges and lots of stuff could be improved, but the basic OpenAPI integration works pretty decently.
